equinox_models/RNN.py

```diff
@@ -0,0 +1,105 @@
+import abc
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+
+
+class _AbstractRNNCell(eqx.Module):
+    """Abstract RNN Cell class."""
+
+    cell: eqx.Module
```

Comment:
Wrapping this in AbstractVar throws an error when you attempt to subclass and give the cell a more specific type annotation, i.e.
```
class LinearCell(AbstractRNNCell):
    cell: eqx.nn.Linear
```
A resolution to this is having all the subclasses use eqx.Module as their type annotations for the cell, what do you think?

Reply:
bleh, I was hoping this wouldn't be an issue since we're not using Pyright. What exactly is triggering this error?

In Python, this technically isn't type safe unless you use properties (overly cumbersome imo) because we can't mark the cell as immutable after initialisation, and subclassed mutable attributes should be type invariant. In this case, we know they are immutable after init, so it's not actually an issue.

A hacky way to get around this is to define `AbstractVar: TypeAlias = getattr(sys, "foo" + "bar")`, which causes the type-checker to give up. This itself is technically incorrect, but see if you can't just tell the type-checker to ignore it.

Reply:
It was actually an equinox TypeError. However, turns out this was a bug which was fixed in the latest update of equinox. When using equinox 0.11.0 there is no error with this change so I have made it now.

Pycharm will give me a `Cannot assign to class variable 'cell' via instance` warning, but passes all pre-commit checks and doesn't give an error when running the model.